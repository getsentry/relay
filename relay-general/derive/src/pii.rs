use proc_macro2::TokenStream;
use quote::quote;

use syn::{Lit, LitBool, Meta};

use crate::{is_newtype, parse_field_name_from_field_attributes};

fn parse_attributes(bi_ast: &syn::Field) -> Option<bool> {
    for attr in &bi_ast.attrs {
        let meta = match attr.interpret_meta() {
            Some(meta) => meta,
            None => continue,
        };

        if meta.name() != "should_strip_pii" {
            continue;
        }

        let name_value = match meta {
            Meta::NameValue(x) => x,
            _ => panic!("Invalid usage of should_strip_pii, need NameValue"),
        };

        match name_value.lit {
            Lit::Bool(LitBool { value, .. }) => return Some(value),
            _ => panic!("Invalid usage of should_strip_pii, need bool as value"),
        };
    }

    None
}

pub fn derive_pii(mut s: synstructure::Structure<'_>) -> TokenStream {
    s.add_bounds(synstructure::AddBounds::None);

    let arms = s.each_variant(|variant| {
        if is_newtype(variant) {
            let inner_ident = &variant.bindings()[0].binding;
            return quote!(crate::processor::Attributes::get_attrs(#inner_ident));
        }

        let mut whitelist = quote!();
        let mut blacklist = quote!();

        for (i, bi) in variant.bindings().iter().enumerate() {
            if let Some(val) = parse_attributes(bi.ast()) {
                let field_name = parse_field_name_from_field_attributes(bi.ast(), i).1;

                if val {
                    whitelist = quote!(#field_name, #whitelist);
                } else {
                    blacklist = quote!(#field_name, #blacklist);
                }
            }
        }

        quote! {
            crate::pii::PiiAttrs {
                whitelist: &[#whitelist],
                blacklist: &[#blacklist],
            }
        }
    });

    s.gen_impl(quote! {
        #[automatically_derived]
        gen impl crate::processor::Attributes<crate::pii::PiiAttrs> for @Self {
            fn get_attrs(&self) -> crate::pii::PiiAttrs {
                match *self {
                    #arms
                }
            }
        }
    })
}
