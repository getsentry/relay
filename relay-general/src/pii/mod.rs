//! PII stripping processor.

mod attachments;
mod builtin;
mod compiledconfig;
mod config;
mod convert;
mod generate_selectors;
mod legacy;
mod minidumps;
mod processor;
mod redactions;
mod regexes;
mod utils;

pub use self::attachments::PiiAttachmentsProcessor;
pub use self::builtin::BUILTIN_RULES;
pub use self::compiledconfig::CompiledPiiConfig;
pub use self::config::{
    AliasRule, MultipleRule, Pattern, PatternRule, PiiConfig, RedactPairRule, RuleSpec, RuleType,
    Vars,
};
pub use self::generate_selectors::selector_suggestions_from_value;
pub use self::legacy::DataScrubbingConfig;
pub use self::minidumps::ScrubMinidumpError;
pub use self::processor::PiiProcessor;
pub use self::redactions::{Redaction, ReplaceRedaction};
