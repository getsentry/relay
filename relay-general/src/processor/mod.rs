//! Provides support for processing structures.

mod attrs;
mod chunks;
mod funcs;
mod impls;
mod selector;
mod size;
mod traits;

pub use self::attrs::{
    BagSize, FieldAttrs, MaxChars, Path, Pii, ProcessingState, UnknownValueTypeError, ValueType,
};
pub use self::chunks::{join_chunks, process_chunked_value, split_chunks, Chunk};
pub use self::funcs::process_value;
pub use self::selector::{SelectorPathItem, SelectorSpec};
pub use self::size::{estimate_size, estimate_size_flat};
pub use self::traits::{ProcessValue, Processor};
