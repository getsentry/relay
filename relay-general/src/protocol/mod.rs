//! Implements the sentry event protocol.

// FIXME: Workaround for https://github.com/GREsau/schemars/pull/65
#![allow(clippy::field_reassign_with_default)]

mod breadcrumb;
mod clientsdk;
mod constants;
mod contexts;
mod debugmeta;
mod event;
mod exception;
mod fingerprint;
mod logentry;
mod measurements;
mod mechanism;
mod metrics;
mod request;
#[cfg(feature = "jsonschema")]
mod schema;
mod security_report;
mod session;
mod span;
mod stacktrace;
mod tags;
mod templateinfo;
mod thread;
mod types;
mod user;
mod user_report;

pub use self::breadcrumb::Breadcrumb;
pub use self::clientsdk::{ClientSdkInfo, ClientSdkPackage};
pub use self::constants::{INVALID_ENVIRONMENTS, INVALID_RELEASES, VALID_PLATFORMS};
pub use self::contexts::{
    AppContext, BrowserContext, Context, ContextInner, Contexts, DeviceContext, GpuContext,
    OperationType, OsContext, RuntimeContext, SpanId, SpanStatus, TraceContext, TraceId,
};
pub use self::debugmeta::{
    AppleDebugImage, CodeId, DebugId, DebugImage, DebugMeta, NativeDebugImage, NativeImagePath,
    SystemSdkInfo,
};
pub use self::event::{
    Event, EventId, EventProcessingError, EventType, ExtraValue, GroupingConfig,
    ParseEventTypeError,
};
pub use self::exception::Exception;
pub use self::fingerprint::Fingerprint;
pub use self::logentry::{LogEntry, Message};
pub use self::measurements::Measurements;
pub use self::mechanism::{CError, MachException, Mechanism, MechanismMeta, PosixSignal};
pub use self::metrics::Metrics;
pub use self::request::{Cookies, HeaderName, HeaderValue, Headers, Query, Request};
#[cfg(feature = "jsonschema")]
pub use self::schema::event_json_schema;
pub use self::security_report::{Csp, ExpectCt, ExpectStaple, Hpkp, SecurityReportType};
pub use self::session::{
    ParseSessionStatusError, SessionAggregateItem, SessionAggregates, SessionAttributes,
    SessionStatus, SessionUpdate,
};
pub use self::span::Span;
pub use self::stacktrace::{Frame, FrameData, FrameVars, RawStacktrace, Stacktrace};
pub use self::tags::{TagEntry, Tags};
pub use self::templateinfo::TemplateInfo;
pub use self::thread::{Thread, ThreadId};
pub use self::types::{
    datetime_to_timestamp, Addr, AsPair, InvalidRegVal, IpAddr, JsonLenientString, LenientString,
    Level, PairList, ParseLevelError, RegVal, Timestamp, Values,
};
pub use self::user::{Geo, User};
pub use self::user_report::UserReport;
