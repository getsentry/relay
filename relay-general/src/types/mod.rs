//! Basic types for the meta protocol.
//!
//! This only provides the types (or aliases) that are used for the meta
//! part of the protocol.  This is the core annotation system as well as
//! the dynamic value parts and the metadata that goes with it.

mod annotated;
mod impls;
mod meta;
mod traits;
mod value;

pub use self::annotated::{
    Annotated, MetaMap, MetaTree, ProcessingAction, ProcessingResult, SerializableAnnotated,
};
pub use self::impls::SerializePayload;
pub use self::meta::{Error, ErrorKind, Meta, Range, Remark, RemarkType};
pub use self::traits::{Empty, FromValue, SkipSerialization, ToValue};
pub use self::value::{to_value, Array, Map, Object, Timestamp, Value, ValueDescription};
