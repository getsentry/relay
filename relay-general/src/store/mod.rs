//! Utility code for sentry's internal store.
use std::sync::Arc;

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::processor::{ProcessingState, Processor};
use crate::protocol::{Event, IpAddr};
use crate::types::{Meta, ProcessingResult};

mod event_error;
mod geo;
mod legacy;
mod normalize;
mod remove_other;
mod schema;
mod transactions;
mod trimming;

pub use crate::store::geo::{GeoIpError, GeoIpLookup};

/// The config for store.
#[derive(Serialize, Deserialize, Debug, Default)]
#[serde(default)]
pub struct StoreConfig {
    pub project_id: Option<u64>,
    pub client_ip: Option<IpAddr>,
    pub client: Option<String>,
    pub key_id: Option<String>,
    pub protocol_version: Option<String>,
    pub grouping_config: Option<Value>,
    pub user_agent: Option<String>,

    pub max_secs_in_future: Option<i64>,
    pub max_secs_in_past: Option<i64>,
    pub enable_trimming: Option<bool>,

    /// When `true`, it is assumed the input already ran through normalization with
    /// is_renormalize=false. `None` equals false.
    pub is_renormalize: Option<bool>,

    /// Overrides the default flag for other removal.
    pub remove_other: Option<bool>,

    /// When `true` it adds context information extracted from the user agent
    pub normalize_user_agent: Option<bool>,

    /// When the event has been sent, according to the SDK. Passed in via envelope headers.
    pub sent_at: Option<DateTime<Utc>>,
}

/// The processor that normalizes events for store.
pub struct StoreProcessor<'a> {
    config: Arc<StoreConfig>,
    normalize: normalize::NormalizeProcessor<'a>,
}

impl<'a> StoreProcessor<'a> {
    /// Creates a new normalization processor.
    pub fn new(config: StoreConfig, geoip_lookup: Option<&'a GeoIpLookup>) -> Self {
        let config = Arc::new(config);
        StoreProcessor {
            normalize: normalize::NormalizeProcessor::new(config.clone(), geoip_lookup),
            config,
        }
    }

    /// Returns a reference to the config.
    pub fn config(&self) -> &StoreConfig {
        &self.config
    }
}

impl<'a> Processor for StoreProcessor<'a> {
    fn process_event(
        &mut self,
        event: &mut Event,
        meta: &mut Meta,
        state: &ProcessingState<'_>,
    ) -> ProcessingResult {
        let is_renormalize = self.config.is_renormalize.unwrap_or(false);
        let remove_other = self.config.remove_other.unwrap_or(!is_renormalize);
        let enable_trimming = self.config.enable_trimming.unwrap_or(true);

        // Convert legacy data structures to current format
        legacy::LegacyProcessor.process_event(event, meta, state)?;

        if !is_renormalize {
            // internally noops for non-transaction events
            // TODO: Parts of this processor should probably be a filter once Relay is store so we
            // can revert some changes to ProcessingAction
            transactions::TransactionsProcessor::new(self.config.sent_at)
                .process_event(event, meta, state)?;
        }

        if !is_renormalize {
            // Check for required and non-empty values
            schema::SchemaProcessor.process_event(event, meta, state)?;

            // Normalize data in all interfaces
            self.normalize.process_event(event, meta, state)?;
        }

        if remove_other {
            // Remove unknown attributes at every level
            remove_other::RemoveOtherProcessor.process_event(event, meta, state)?;
        }

        if !is_renormalize {
            // Add event errors for top-level keys
            event_error::EmitEventErrors::new().process_event(event, meta, state)?;
        }

        if enable_trimming {
            // Trim large strings and databags down
            trimming::TrimmingProcessor::new().process_event(event, meta, state)?;
        }

        Ok(())
    }
}
