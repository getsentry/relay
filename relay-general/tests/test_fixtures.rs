use std::fs;

use relay_general::pii::{PiiConfig, PiiProcessor};
use relay_general::processor::{process_value, ProcessingState};
use relay_general::protocol::Event;
use relay_general::store::{StoreConfig, StoreProcessor};
use relay_general::types::{Annotated, SerializableAnnotated};

use insta::assert_yaml_snapshot;

macro_rules! event_snapshot {
    ($id:ident) => {
        mod $id {
            use super::*;

            lazy_static::lazy_static! {
                static ref PII_CONFIG: PiiConfig = PiiConfig::from_json(r##"{
                  "rules": {
                    "removeOkDetectToken": {
                      "type": "pattern",
                      "pattern": ".token.:.([0-9]+).",
                      "redaction": {"method": "replace", "text": "[ok detect token]"}
                    },
                    "removeDotLocal": {
                      "type": "pattern",
                      "pattern": "(.*)[.]local",
                      "redaction": {"method": "replace", "text": "[.local hostname]"}
                    }
                  },
                  "applications": {
                    "tags.server_name": ["removeDotLocal"],
                    "tags.RequestId": ["@anything:remove"],

                    "contexts.device.boot_time": ["@anything:remove"],
                    "contexts.device.screen_resolution": ["@anything:remove"],
                    "contexts.device.screen_density": ["@anything:remove"],
                    "contexts.device.screen_height_pixels": ["@anything:remove"],
                    "contexts.device.screen_width_pixels": ["@anything:remove"],
                    "contexts.device.screen_dpi": ["@anything:remove"],
                    "contexts.device.memory_size": ["@anything:remove"],
                    "contexts.device.timezone": ["@anything:remove"],

                    "user.ip_address": ["@anything:remove"],
                    "user.email": ["@anything:hash"],

                    "request.cookies.wcsid": ["@anything:replace"],
                    "request.cookies.hblid": ["@anything:replace"],
                    "request.cookies._okdetect": ["removeOkDetectToken"],
                    "request.headers.MS-ASPNETCORE-TOKEN": ["@anything:replace"],
                    "request.headers.User-Agent": ["@anything:replace"],

                    "request.env.DOCUMENT_ROOT": ["@userpath:replace"],
                    "exception.values.*.stacktrace.frames.*.filename": ["@userpath:replace"],
                    "exception.values.*.stacktrace.frames.*.abs_path": ["@userpath:replace"],
                    "breadcrumbs.values.*.message": ["@userpath:replace"]
                  }
                }"##).unwrap();
            }

            fn load_fixture() -> Annotated<Event> {
                let data = fs::read_to_string(
                    format!("tests/fixtures/payloads/{}.json", stringify!($id))
                ).unwrap();
                Annotated::<Event>::from_json(&data).unwrap()
            }

            #[test]
            fn test_processing() {
                let mut event = load_fixture();

                let config = StoreConfig::default();
                let mut processor = StoreProcessor::new(config, None);
                process_value(&mut event, &mut processor, ProcessingState::root()).unwrap();

                let mut processor = PiiProcessor::new(&*PII_CONFIG);
                process_value(&mut event, &mut processor, ProcessingState::root()).unwrap();
                assert_yaml_snapshot!("pii_stripping", SerializableAnnotated(&event), {
                    ".received" => "[received]",
                    ".timestamp" => "[timestamp]"
                });
            }
        }
    }
}

event_snapshot!(android);
event_snapshot!(cocoa);
event_snapshot!(cordova);
event_snapshot!(dotnet);
event_snapshot!(legacy_python);
event_snapshot!(legacy_node_exception);
