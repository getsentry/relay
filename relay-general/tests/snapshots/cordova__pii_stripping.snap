---
source: relay-general/tests/test_fixtures.rs
expression: SerializableAnnotated(&event)
---
event_id: ae3a2cc6e08741c5b09946c65c022469
level: fatal
type: error
transaction: main at EventEmitter.ipcMain.on
logentry:
  formatted: "Error: Error triggered in main processes"
logger: ""
platform: other
timestamp: "[timestamp]"
received: "[received]"
release: sentry-electron-test@2.0.2
environment: development
user:
  ip_address: ~
contexts:
  app:
    app_name: Sentry Crash Test
    app_version: 2.0.2
    type: app
  chrome:
    name: Chrome
    version: 61.0.3163.100
    type: runtime
  device:
    arch: x64
    type: device
  node:
    name: Node
    version: 8.9.3
    type: runtime
  os:
    name: Mac OS X
    version: 10.13.6
    build: 17G65
    kernel_version: 17.7.0
    type: os
  runtime:
    name: Electron
    version: 2.0.2
    type: runtime
breadcrumbs:
  values:
    - timestamp: 1532692006.665
      type: ui
      category: electron
      level: info
      message: app.will-finish-launching
    - timestamp: 1532692006.715
      type: ui
      category: electron
      level: info
      message: app.ready
    - timestamp: 1532692006.734
      type: ui
      category: electron
      level: info
      message: app.session-created
    - timestamp: 1532692006.754
      type: ui
      category: electron
      level: info
      message: app.web-contents-created
    - timestamp: 1532692006.796
      type: ui
      category: electron
      level: info
      message: app.browser-window-created
    - timestamp: 1532692006.918
      type: ui
      category: electron
      level: info
      message: app.browser-window-focus
    - timestamp: 1532692007.165
      type: ui
      category: electron
      level: info
      message: "WebContents[1].dom-ready"
    - timestamp: 1532692011.34
      type: default
      category: ui.click
      level: info
      message: body > div.main > div.group > a.crash
exception:
  values:
    - type: Error
      value: Error triggered in main processes
      stacktrace:
        frames:
          - function: "WebContents.emitter.emit.args [as emit]"
            module: backend
            filename: "/Users/[user]/Projects/sentry-electron/dist/main/backend.js"
            abs_path: "/Users/[user]/Projects/sentry-electron/dist/main/backend.js"
            lineno: 278
            colno: 20
            pre_context:
              - "                    category: 'electron',"
              - "                    message: `${category}.${event}`,"
              - "                    timestamp: new Date().getTime() / 1000,"
              - "                    type: 'ui',"
              - "                };"
              - "                minimal_1.addBreadcrumb(breadcrumb);"
              - "            }"
            context_line: "            return emit(event, ...args);"
            post_context:
              - "        };"
              - "    }"
              - "    /** Loads new native crashes from disk and sends them to Sentry. */"
              - "    sendNativeCrashes(extra) {"
              - "        return __awaiter(this, void 0, void 0, function* () {"
              - "            // Whenever we are called, assume that the crashes we are going to load down"
              - "            // below have occurred recently. This means, we can use the same event data"
            in_app: false
          - function: WebContents.emit
            module: events
            filename: events.js
            abs_path: events.js
            lineno: 214
            colno: 7
            in_app: false
          - function: emitTwo
            module: events
            filename: events.js
            abs_path: events.js
            lineno: 126
            colno: 13
            in_app: false
          - function: WebContents.<anonymous>
            module: "electron.dist.Electron.app.Contents.Resources.electron.asar.browser.api:web-contents"
            filename: "app:///node_modules/electron/dist/Electron.app/Contents/Resources/electron.asar/browser/api/web-contents.js"
            abs_path: "app:///node_modules/electron/dist/Electron.app/Contents/Resources/electron.asar/browser/api/web-contents.js"
            lineno: 286
            colno: 13
            pre_context:
              - ""
              - "  // Every remote callback from renderer process would add a listenter to the"
              - "  // render-view-deleted event, so ignore the listenters warning."
              - "  this.setMaxListeners(0)"
              - ""
              - "  // Dispatch IPC messages to the ipc module."
              - "  this.on('ipc-message', function (event, [channel, ...args]) {"
            context_line: "    ipcMain.emit(channel, event, ...args)"
            post_context:
              - "  })"
              - "  this.on('ipc-message-sync', function (event, [channel, ...args]) {"
              - "    Object.defineProperty(event, 'returnValue', {"
              - "      set: function (value) {"
              - "        return event.sendReply(JSON.stringify(value))"
              - "      },"
              - "      get: function () {}"
            in_app: false
          - function: EventEmitter.emit
            module: events
            filename: events.js
            abs_path: events.js
            lineno: 211
            colno: 7
            in_app: false
          - function: emitOne
            module: events
            filename: events.js
            abs_path: events.js
            lineno: 116
            colno: 13
            in_app: false
          - function: EventEmitter.ipcMain.on
            module: main
            filename: "app:///main.js"
            abs_path: "app:///main.js"
            lineno: 38
            colno: 9
            pre_context:
              - ""
              - "app.on('window-all-closed', () => app.quit());"
              - ""
              - // The IPC handlers below trigger errors in the here (main process) when
              - // the user clicks on corresponding buttons in the UI (renderer).
              - "ipcMain.on('demo.error', () => {"
              - "  console.log('Error triggered in main processes');"
            context_line: "  throw new Error('Error triggered in main processes');"
            post_context:
              - "});"
              - ""
              - "ipcMain.on('demo.crash', () => {"
              - "  console.log('process.crash()');"
              - "  process.crash();"
              - "});"
              - ""
            in_app: true
tags:
  - - event_type
    - javascript
extra:
  crashed_process: browser
sdk:
  name: sentry.javascript.electron
  version: 0.8.0
  packages:
    - name: "npm:@sentry/node"
      version: 4.0.0-beta.11
    - name: "npm:@sentry/node"
      version: 0.8.0
_meta:
  exception:
    values:
      "0":
        stacktrace:
          frames:
            "0":
              abs_path:
                "":
                  rem:
                    - - "@userpath:replace"
                      - s
                      - 7
                      - 13
                  len: 46
              filename:
                "":
                  rem:
                    - - "@userpath:replace"
                      - s
                      - 7
                      - 13
                  len: 46
  user:
    ip_address:
      "":
        rem:
          - - "@anything:remove"
            - x
